volumebinding: scheduler queueing hints - CSIDriver

YamasouA opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
kube-scheduler implements scheduling hints for the VolumeBinding plugin.
The scheduling hints allow the scheduler to retry scheduling a Pod that was previously rejected by the VolumeBinding plugin only if a new resource referenced by the plugin was created or an existing resource referenced by the plugin was updated.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes/kubernetes/issues/118893

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The scheduler implements QueueingHint in VolumeBinding plugin's CSIDriver event, which enhances the throughput of scheduling.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4247
```

YamasouA commented:
/cc @sanposhiho @carlory

sanposhiho reviewed (commented):
/kind feature
/ok-to-test

sanposhiho commented:
/retest

sanposhiho commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -117,12 +119,39 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
 		// When CSIStorageCapacity is enabled, pods may become schedulable
 		// on CSI driver & storage capacity changes.
-		{Event: framework.ClusterEvent{Resource: framework.CSIDriver, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.CSIDriver, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterCSIDriverChange},
```
@carlory 

Is there any scenario where CSIDriver creation makes Pod schedulable? We skip a capacity check when it's not found.
https://github.com/kubernetes/kubernetes/blob/8565e375251450a291f0af3c6195c7a5bf890292/pkg/scheduler/framework/plugins/volumebinding/binder.go#L1014-L1018

sanposhiho replied:
cc @kubernetes/sig-storage-misc 

---

/hold 
QHint implementation depends on whether we really need Creation event or not.

carlory replied:
It's introduced by https://github.com/kubernetes/kubernetes/pull/100003. 

Let's the primary reviewer of that PR to decide whether to remove the ADD event or not.

/cc @Huang-Wei @ahg-g 

IMO, the ADD event is unnecessary as you said. 

/cc @jsafrane @xing-yang 

Can you also take another look?

jsafrane replied:
Good catch, I agree that creation of a CSIDriver cannot make a pending pod schedulable and we don't need Add event here.

This is the only place where CSIDriver instances are checked:
https://github.com/kubernetes/kubernetes/blob/d589a3f866144087e19d092a63b2ef275f2ab912/pkg/scheduler/framework/plugins/volumebinding/binder.go#L1013-L1022
A newly created CSIDriver can make a pod un-schedulable there, but not schedulable.

carlory replied:
@YamasouA please remove the `ADD` event.

YamasouA replied:
@jsafrane @carlory 
Thankyou for your review!
I removed ADD event!!

sanposhiho replied:
/unhold

sanposhiho commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -117,12 +119,39 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
+			continue
+		}
+		driverSpec := modifiedCSIDriver.Spec
+		if driverSpec.StorageCapacity != nil && *driverSpec.StorageCapacity {
```
Is it the opposite?

```suggestion
		if driverSpec.StorageCapacity == nil || !*driverSpec.StorageCapacity {
```

---

When `driver.Spec.StorageCapacity == nil || !*driver.Spec.StorageCapacity`, we return true on `hasEnoughCapacity`.
https://github.com/kubernetes/kubernetes/blob/8565e375251450a291f0af3c6195c7a5bf890292/pkg/scheduler/framework/plugins/volumebinding/binder.go#L1022-L1024

Meaning, if driver's `StorageCapacity` is modified to match `driver.Spec.StorageCapacity == nil || !*driver.Spec.StorageCapacity`, the modification could make Pod schedulable, right?

YamasouA replied:
@sanposhiho 
I apologize for my misunderstanding!
I have corrected the issues.[e225283](https://github.com/kubernetes/kubernetes/pull/125171/commits/e225283f9d20284d9f404abfa412ebbc37f44e0c)

YamasouA commented:
/retest

sanposhiho commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -117,12 +119,39 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
+		}
+	}
+
+	logger.V(5).Info("CISDriver was created or updated but it doesn't make this pod schedulable.")
```
```suggestion
	logger.V(5).Info("CSIDriver was created or updated but it doesn't make this pod schedulable")
```

sanposhiho commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -117,12 +119,39 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
+		if vol.CSI == nil || vol.CSI.Driver != modifiedCSIDriver.Name {
+			continue
+		}
+		driverSpec := modifiedCSIDriver.Spec
+		if driverSpec.StorageCapacity == nil || !*driverSpec.StorageCapacity {
+			logger.V(5).Info("CSIDriver was created or updated and this may make the Pod rejected by VolumeBinding plugin in the previous scheduling cycle schedulable")
+			return framework.Queue, err
+		}
```
```suggestion
		if (originalCSIDriver.Spec.StorageCapacity != nil && *originalCSIDriver.Spec.StorageCapacity) && (modifiedCSIDriver.Spec.StorageCapacity == nil || !*modifiedCSIDriver.Spec.StorageCapacity) {
			logger.V(5).Info("CSIDriver was updated and storage capacity got disabled, which may make the pod schedulable")
			return framework.Queue, err
		}
```

YamasouA replied:
I fix this comment
[31ec811](https://github.com/kubernetes/kubernetes/pull/125171/commits/31ec811546b4e8e2dd202838a3b0857dedc53781)

carlory commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -117,12 +119,39 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
+		if (originalCSIDriver.Spec.StorageCapacity != nil && *originalCSIDriver.Spec.StorageCapacity) &&
+			(modifiedCSIDriver.Spec.StorageCapacity == nil || !*modifiedCSIDriver.Spec.StorageCapacity) {
+			logger.V(5).Info("CSIDriver was updated and storage capacity got disabled, which may make the pod schedulable")
+			return framework.Queue, err
```
```suggestion
			return framework.Queue, nil
```

YamasouA replied:
@carlory 
I fix this comment!
[58eb993](https://github.com/kubernetes/kubernetes/pull/125171/commits/58eb9937d5063a3cd793f07bd867d786a6475cc7)

sanposhiho commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -117,12 +119,39 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
+		}
+	}
+
+	logger.V(5).Info("CISDriver was created or updated but it doesn't make this pod schedulable")
```
```suggestion
	logger.V(5).Info("CSIDriver was created or updated but it doesn't make this pod schedulable")
```

sanposhiho commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding_test.go:
```diff
@@ -890,3 +890,190 @@ func TestVolumeBinding(t *testing.T) {
+			expect: framework.Queue,
+		},
+		{
+			name: "driver name in pod and csidriver name are different.",
```
```suggestion
			name: "driver name in pod and csidriver name are different",
```

sanposhiho commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding_test.go:
```diff
@@ -890,3 +890,190 @@ func TestVolumeBinding(t *testing.T) {
+					Name: "test1",
+				},
+				Spec: storagev1.CSIDriverSpec{
+					StorageCapacity: &trueVar,
```
Instead of `trueVar` and `falseVar`, we can use `ptr.To(true)` and `ptr.To(false)`.

sanposhiho commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding_test.go:
```diff
@@ -890,3 +890,190 @@ func TestVolumeBinding(t *testing.T) {
+					Name: "test2",
+				},
+				Spec: storagev1.CSIDriverSpec{
+					StorageCapacity: &trueVar,
```
So, we have to test these two behaviours:
- [First] When a driver name in the pod and a csidriver name are different, it returns Skip.
- [Second] When a csidriver's storage capacity is still enabled after the update, it returns Skip.

And, you want to make sure the first one with this test case. Then, it's better to change like ↓ so that this test case returns Skip for the first point, not for the second point.

```suggestion
					StorageCapacity: ptr.To(false),
```

Then, if there was a bug for the first point, this test case would fail.

YamasouA replied:
I fix it as per your comment.
[f46d8e9](https://github.com/kubernetes/kubernetes/pull/125171/commits/f46d8e9c2abbfad2e46270864ed150aa46763aa7)

sanposhiho reviewed (commented):
/lgtm

sanposhiho commented:
/assign @saad-ali

For approval. Also, please check https://github.com/kubernetes/kubernetes/pull/125171#discussion_r1622231476, which isn't yet concluded from SIG Storage side too.

sanposhiho commented:
@jsafrane Can you possibly take a look at this one too? Another QHint implementation.

jsafrane commented:
Oops, sorry, I missed this one. Please rebase, but I am afraid we will miss 1.31.

YamasouA committed:
volumebinding: scheduler queueing hints - CSIDriver

fix if condition

add test

add log

eliminate unnecessary args from log

fix Queue condition

check original pod status

fix return value when can scheduleable

fix tweak

fix testcase

YamasouA commented:
@sanposhiho @jsafrane
I made a mistake with the rebase operation, and the commit log has become messy. I apologize for the inconvenience. Also, the LGTM has been removed, so could you please review it again? Thank you.

sanposhiho commented:
@YamasouA I'd suggest you clean up labels that the bot added when you made a mistake rebase.

sanposhiho commented:
also can you squash the commits?

sanposhiho commented:
/retest

YamasouA commented:
/remove-area apiserver
/remove-area cloudprovider
/remove-area code-generation
/remove-area conformance
/remove-area dependency
/remove-area e2e-test-framework
/remove-area ipvs
/remove-area kube-proxy
/remove-area kube-adm
/remove-area kubectl
/remove-area kubelet
/remove-area provider/gcp
/remove-area release-end
/remove-area test

YamasouA commented:
/remove-area kubeadm
/remove-area release-eng
/remove-sig api-machinery
/remove-sig apps
/remove-sig arhitecture
/remove-sig auth
/remove-sig autoscalling
/remove-sig cli
/remove-sig cloud-provider
/remove-sig cluster-lifecycle
/remove-sig etcd
/remove-sig instrumentation
/remove-sig network
/remove-sig node
/remove-sig release
/remove-sig scalability
/remove-sig testing
/remove-sig windows

YamasouA commented:
/remove-sig autoscaling
/remove-sig architecture
/remove-kind api-change

sanposhiho reviewed (commented):
/lgtm

For `/approve`, we need the conclusion of -
https://github.com/kubernetes/kubernetes/pull/125171/files#r1625479464

sanposhiho commented:
@jsafrane Would it be possible to take a look?

YamasouA committed:
delete framework.Add

YamasouA committed:
tweak

YamasouA commented:
/retest

carlory commented:
@YamasouA Please try to rebase upstream and re-run the failing test. If the test still fails, you have to fix the test.

sanposhiho commented:
@YamasouA Can you take a look at the CI failure?

YamasouA committed:
fix unit test

YamasouA commented:
@carlory @sanposhiho 
Sorry for the delay, and thank you for your support.
I fixed the tests by referencing @carlory 's PR.

YamasouA commented:
Even though I'm following upstream like this, why are conflicts occurring in the tests?
```
git fetch upstream
git rebase upstream/master
```

YamasouA committed:
tweak

carlory commented:
/retest

carlory commented:
please eidt the description of this PR. The format is wrong.

YamasouA commented:
@carlory 
Thankyou for your advice.

carlory commented:
/lgtm

For storage part

carlory commented:
/cc @sanposhiho

sanposhiho commented:
/cc @macsko @dom4ha

YamasouA committed:
fix test

YamasouA commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding_test.go:
```diff
@@ -1372,3 +1373,188 @@ func TestIsSchedulableAfterCSIStorageCapacityChange(t *testing.T) {
+			},
+			err:    false,
+			expect: framework.QueueSkip,
+		},
```
@macsko 
Thankyou for your review.
I fix this comment.
Please review again.

sanposhiho commented:
/retest

sanposhiho reviewed (approved):
/approve

Leave `/lgtm` to @carlory or @macsko.

sanposhiho commented:
/approve

macsko commented:
/lgtm

sanposhiho commented:
/retest

sanposhiho commented:
/retest

sanposhiho commented:
/retest

(I'm `/retest`ing the integ test so many times, but not related for sure.)
